packages/design-system/src/components/Button/Primitive/ButtonPrimitive.tsx

```diff
@@ -1,78 +1,84 @@
-import React, { forwardRef, Ref } from 'react';
+import React, { forwardRef, Ref, ReactElement } from 'react';
 // eslint-disable-next-line @talend/import-depth
-import { IconName } from '@talend/icons/dist/typeUtils';
+import { IconNameWithSize } from '@talend/icons/dist/typeUtils';
 import classnames from 'classnames';
 import Clickable, { ClickableProps } from '../../Clickable';
 
+import { DataAttributes, DeprecatedIconNames } from '../../../types';
 import { StackHorizontal } from '../../Stack';
 import Loading from '../../Loading';
-import { Icon } from '../../Icon/Icon';
-import { DataAttributes } from '../../../types';
+import { parseDeprecatedIcon } from '../../Icon/DeprecatedIconHelper';
 
 import styles from './ButtonStyles.module.scss';
+import { SizedIcon } from '../../Icon';
 
 export type AvailableVariantsTypes = 'primary' | 'destructive' | 'secondary' | 'tertiary';
 export type AvailableSizes = 'M' | 'S';
 export type ButtonVariantType<T extends AvailableVariantsTypes, P extends object> = {
 	variant: T;
 } & P;
 
-export type SharedButtonTypes = {
-	size?: AvailableSizes;
-	icon?: IconName | React.ReactElement;
+export type SharedButtonTypes<S extends AvailableSizes> = {
 	isLoading?: boolean;
 	isDropdown?: boolean;
+	size?: S;
+	icon?: S extends 'S'
+		? IconNameWithSize<'S'>
+		: DeprecatedIconNames | ReactElement | IconNameWithSize<'M'>;
```

Comment:
For legacy reasons, we still have to support the rest. This translate to `ReactElement | string` since `DeprecatedIconNames` is a union of string with `string`. 

Historically this was done so icons could be fetched through passing the name `remote-balbla` or even images by passing `src-blabla`.

I would love to deprecate this but it would be a large breaking change.